# Clean Up Your Game: Tackling a Messy Football Dataset

Playing FIFA 21 is not just a game, it's a passion. And for passionate gamers and football fans, keeping track of player stats and match results is essential. But what happens when the data in the FIFA 21 game dataset is messy and incomplete? Enter our FIFA 21 cleaning dataset, designed to help you clean and organize your FIFA 21 data with ease. With this dataset, you can say goodbye to missing values, inconsistent formatting, and other data cleaning headaches. So whether you're a die-hard fan or a data analyst, get ready to take your FIFA 21 data to the next level with our FIFA 21 cleaning dataset.

### 1. Data Assessment

In [1]:
# Import necessary libraries
import pandas as pd
import numpy as np
from unidecode import unidecode
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

This code  imports several Python modules that are commonly used for data manipulation and visualization.

- **pandas** and **numpy** are used for data manipulation and analysis, providing data structures and functions needed to work with structured data seamlessly.

- **unidecode** is used for converting Unicode text to ASCII text, which is useful when working with non-English text.

- **seaborn** and **matplotlib** are used for data visualization, providing a range of tools for creating plots, charts, and graphs.

- **warnings**: A module used for issuing warning messages in Python.

Overall, this code sets up the necessary environment for data analysis and visualization in Python.

In [2]:
df = pd.read_csv('./data/fifa21 raw data v2.csv')
df.head()

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28    91   

   POT                         Club  ...     A/W     D/W   IR PAC SHO  PAS  \
0   93         \n\n\n\nFC Barcelona  ...  Medium     Low  5 ★  85  92   91   
1   92             \n\n\n\nJuventus  ...    High     Low  5 ★  89  93   81   
2   93      \n\n\n\nAtlético Madrid  ...  Medium  Medium  3 ★  87  92   78   
3   91      \n\n\n\nManchester City  ...    High    High  4 ★  76  86   93   
4   91  \n\n\n\nParis Saint-Germain  ...    High  Medium  5 ★  91  85   86   

  DRI DEF PHY Hits  
0  95  38  65  771  
1  89  35  77  562  
2  90  52  90  150  
3  88  64  78  207  
4  94  36  59  595  

[5 rows x 77 columns]

We start by reading in a FIFA 21 raw data CSV file using the **pd.read_csv()** function from the **pandas** library. The file is located in the **./data/** directory and is named **fifa21 raw data v2.csv**. The resulting DataFrame is assigned to the variable **df**.

Next, we use the **head()** method to display the first 5 rows of the DataFrame. This method is useful for getting a quick look at the structure and content of the DataFrame.

Together, these two lines of code allow us to load and preview the FIFA 21 raw data CSV file in a pandas DataFrame, making it easy to manipulate and analyze the data further.

From the given dataset, we can identify some potential messy data issues, such as:

1. **Missing values**
2. **Inconsistent data format**
3. **Inconsistent naming conventions**
4. **Redundant data**

In [3]:
# Inspect DataFrame
df.shape

(18979, 77)

This output of **df.shape** means that the Pandas DataFrame df has 18979 rows and 77 columns.

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   LongName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   ↓OVA              18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   Club              18979 non-null  object
 10  Contract          18979 non-null  object
 11  Positions         18979 non-null  object
 12  Height            18979 non-null  object
 13  Weight            18979 non-null  object
 14  Preferred Foot    18979 non-null  object
 15  BOV               18979 non-null  int64 
 16  Best Position     18979 non-null  object
 17  Joined      

The **info()** method in Pandas provides a summary of a DataFrame including the number of non-null values, data types, memory usage, and more. It can be used to quickly inspect the structure and content of a DataFrame.

From the output, we can see that there are missing values in the **"Loan Date End"** and **"Hits"** columns as they have fewer non-null values compared to the other columns. This suggests that the data may be messy, and some data cleaning or imputation may be necessary. Also, some columns have data types that may not be suitable for the analysis, such as object data types for numerical data or date/time columns. These columns may need to be converted to the appropriate data type for further analysis.

In [5]:
# Checking for the sum of missing values in each column
df.isna().sum()

ID              0
Name            0
LongName        0
photoUrl        0
playerUrl       0
             ... 
PAS             0
DRI             0
DEF             0
PHY             0
Hits         2595
Length: 77, dtype: int64

The output of **df.isna().sum()** gives the count of missing values (NaN) in each column of the dataframe. In this case, there are no missing values in most of the columns, but there are some missing values in the **"Loan Date End"** and **"Hits"** columns.

In [6]:
# Checking for duplicates in the datasets
df.duplicated().sum()

0

This code will return the number of duplicate rows in the DataFrame **'df'**.

In [7]:
# Create a copy of the DataFrame
df_copy = df.copy()

The code **df_copy = df.copy()** creates a copy of a Pandas DataFrame **df**. This is important because it allows us to make changes to the copy without affecting the original DataFrame.
This is especially important in data analysis and manipulation, where we want to experiment with different transformations and operations on the data without altering the original dataset. By creating a copy, we can keep the original data intact and avoid any unintended consequences of modifying the original data.

Therefore, making a copy of the DataFrame using the **.copy()** method is an important step in many data analysis workflows.


### 2. Data Cleaning

#### 2.1 Dealing with missing values and inconsistent data formats and types

In [8]:
# Removing newline characters(/n) from the Club column
df_copy['Club'] = df_copy['Club'].str.strip()

This line of the code, **df_copy['Club'] = df_copy['Club'].str.strip()**, is using the **strip()** method to remove any leading or trailing whitespace characters from the **"Club"** column. This is useful because sometimes text data can contain extra whitespace characters, such as newline characters **(\n)**, which can make it difficult to work with the data.

The **str** attribute is used to access string methods that can be applied to each element of the **"Club"** column, allowing us to remove any unwanted characters from the strings in the column.

Overall, this code is a useful step in data preprocessing to ensure that the text data in the **"Club"** column is formatted consistently and is easier to work with in downstream analysis.

In [9]:
# Fill missing Hits with 0
df_copy['Hits'] = df_copy['Hits'].fillna(0)

When handling missing data, it is important to carefully consider the context in which the missing values occur in order to determine an appropriate approach for addressing them. In this specific case, the missing values in the **"Hits"** column of the dataset may indicate that certain player profiles have not been clicked on or viewed.

One approach to handling missing data in the **"Hits"** column could be to fill the missing values with zeros. This is a reasonable approach because it is likely that the missing values indicate that certain player profiles have not been viewed or clicked on, and thus would have zero hits. By filling in the missing values with zeros, we are explicitly acknowledging that these profiles have not been viewed, and we can treat them as if they have zero hits for the purposes of analysis.

It is important to note that this approach may not always be appropriate for other datasets or contexts, and different approaches may be needed depending on the nature of the missing data and the goals of the analysis. In general, it is important to carefully consider the context of the missing data and select an appropriate approach that takes into account the potential biases or implications of different strategies for handling missing values.

In [10]:
# Taking care of the inconsistent data format and types in the Hits column
def clean_hits(hits):
    if type(hits) == str:
        if hits[-1]=='K':
            hits = float(hits.replace('K',''))*1000
        return float(hits)
    else:
        return hits

df_copy['Hits'] = df_copy['Hits'].apply(clean_hits).astype('int')

This code is a function **clean_hits** that takes in one argument hits, which is expected to be a string or a number. The function checks if the type of hits is a string. If it is, the function removes the letter **"K"** at the end of the string (if present) and converts the remaining string to a float value. It then multiplies the float value by **1000** to convert it to the corresponding number of hits in thousands. If hits is already a number (not a string), the function simply returns the original value.

The **df_copy['Hits']** column is then transformed using the **apply** method with the clean_hits function to clean up the inconsistent data format and types in the Hits column. The resulting column is then converted to an integer data type using the **astype** method and saved back to the **df_copy['Hits']** column.


In [11]:
# Taking care of the inconsistent data format and types in the 'Loan Date End' and 'Joined' column
df_copy['Loan Date End'] = pd.to_datetime(df_copy['Loan Date End'])
df_copy['Joined'] = pd.to_datetime(df_copy['Joined'])

def year_range(row):
    return  f"{row['Joined'].year} ~ {row['Loan Date End'].year}"

df_copy.loc[~df_copy['Loan Date End'].isna(),'Contract']= df_copy.loc[~df_copy['Loan Date End'].isna()].apply(year_range,axis=1)

df_copy['player_transfer_agreement'] = pd.Series(dtype='str')
df_copy.loc[~df_copy['Loan Date End'].isna(),'player_transfer_agreement'] = 'Loan'
df_copy.loc[df_copy['Loan Date End'].isna(),'player_transfer_agreement'] = 'Contract'
df_copy.loc[df_copy['Contract']=='Free','player_transfer_agreement']='Free Agent'

df_copy[['start_year', 'end_year']] = df_copy['Contract'].str.split('~', expand=True)
df_copy.loc[df_copy['Contract']=='Free',['end_year']]= 0
df_copy.loc[df_copy['Contract']=='Free',['start_year']] = df_copy['Joined'].dt.year
df_copy['start_year'] = df_copy['start_year'].astype('int')
df_copy['end_year'] = df_copy['end_year'].astype('int')
df_copy['contract_duration'] = df_copy['end_year'] - df_copy['start_year']

df_copy.loc[df_copy['Loan Date End'].isna(),'Loan Date End'] = df_copy.apply(lambda row: row['Joined'] + pd.DateOffset(years=row['contract_duration']) if row['contract_duration'] > 0 else pd.NaT, axis=1)

This code is performing several operations on the DataFrame df_copy to extract information related to player transfers and contracts.

* The first part of the code is converting the **'Loan Date End'** and **'Joined'** columns to datetime format, and creating a function **year_range** that concatenates the start and end years of the player's contract.

* Next, a new column **'player_transfer_agreement'** is added to differentiate between **loan agreements** and **regular contracts**, with a third category **'Free Agent'** added for players who are not under contract.

* Then, the start and end years of the player's contract are extracted into separate columns, with a value of **0** added for players who are free agents. The **contract duration** is also calculated and stored in a new column.

* Finally, for players who are currently under contract **(i.e. 'Loan Date End' is NaT)**, the end date is calculated based on the start year and contract duration. If the contract duration is 0, indicating that the player's contract has expired thereby making them free agents, the **'Loan Date End'** is set to **NaT**.

In [12]:
# Taking care of the inconsistent data format and types in the 'Heights' column
def convert_height(height):
    if pd.isnull(height):
        return height
    elif "'" in height:
        feet, inches = height.split("'")
        total_inches = int(feet) * 12 + int(inches.replace('"', ''))
        return int(total_inches * 2.54)
    elif "cm" in height:
        return int(height.replace("cm", ""))
    else:
        return None

df_copy['Height'] = df_copy['Height'].apply(convert_height)

The **convert_height** function is used to convert the values in the **"Height"** column to centimeters. It first checks if the value is null, if it is then it returns the same value. If the value is not null, it checks if the height is in the format of feet and inches or centimeters. If it's in the format of feet and inches, it converts the height to inches and then to centimeters. If it's already in the format of centimeters, it just removes the "cm" and returns the value. If it's not in either of these formats, it returns None.

The function is applied to the **"Height"** column using the apply method, and the converted values are stored back in the same column.

In [13]:
# Taking care of the inconsistent data format and types in the 'Weight' column
def clean_weight(weight):
    if isinstance(weight, float):
        # Convert pounds to kilograms
        return round(weight * 0.45359237, 1)
    elif isinstance(weight, str):
        if weight.endswith("lbs"):
            # Convert pounds to kilograms
            return round(float(weight.strip("lbs")) * 0.45359237, 1)
        elif weight.endswith("kg"):
            # Convert kg to kg
            return round(float(weight.strip("kg")), 1)
        else:
            return None
    else:
        return None

df_copy['Weight'] = df_copy['Weight'].apply(clean_weight)

The code above defines a function called **clean_weight** which takes a weight value as input and returns the weight value in kilograms. The function checks the input value to determine whether it is a string or a float. If the input value is a float, it assumes that the weight is already in kilograms and returns the value as is. If the input value is a string, the function checks whether the string ends with **"lbs"** or **"kg"** and performs the necessary conversion to kilograms. If the string contains feet and inches, the function converts the value to inches and then to kilograms. The function rounds the final result to one decimal place before returning it. Finally, the code applies the **clean_weight** function to the **'Weight'** column.

In [14]:
# Removing the '*' in these columns 'W/F','SM' and 'IR'.
df_copy['W/F'] = df_copy['W/F'].str.replace('★', '')
df_copy['SM'] = df_copy['SM'].str.replace('★', '')
df_copy['IR'] = df_copy['IR'].str.replace('★', '')

These lines of code replace the **'★'** character in the **'W/F'**, **'SM'**, and **'IR'** columns with an empty string, effectively removing it from the column values. This is useful for cases where the **'★'** character is used to indicate a player's skill level or rating, but is not necessary for data analysis or modeling purposes. By removing the character, the column values can be converted to a numerical data type if necessary, without having to deal with the **'★'** character as an additional feature.

In [15]:
# Taking care of the inconsistent data format and types in the 'Release Clause', 'Wage' and 'Value' columns.
def clean_value(value_str):
    if value_str[-1] == 'K':
        return float(value_str[1:-1]) * 1000
    elif value_str[-1] == 'M':
        return float(value_str[1:-1]) * 1000000
    else:
        return float(value_str[1:])

df_copy['Release Clause'] = df_copy['Release Clause'].apply(clean_value).astype('int')
df_copy['Wage'] = df_copy['Wage'].apply(clean_value).astype('int')
df_copy['Value'] = df_copy['Value'].apply(clean_value).astype('int')

This code addresses the inconsistent formatting and types of data in the **'Release Clause'**, **'Wage'**, and **'Value'** columns of the dataset. The function **clean_value** converts the string values to float and multiplies them by **1000** or **1000000**, depending on whether the value ends with **'K'** or **'M'**, respectively. If the value does not end with either of those characters, it is converted to float and returned as is. The apply method is used to apply this function to each value in the three columns, and the astype method is used to convert the resulting values to integers for consistency.

In [16]:
# Converting the necessary columns data type to the category data type
df_copy[['player_transfer_agreement','W/F','SM','Preferred Foot','D/W','A/W','IR']] = df_copy[['player_transfer_agreement','W/F','SM','Preferred Foot','D/W','A/W','IR']].astype('category')

The columns being converted are **"player_transfer_agreement", "W/F", "SM", "Preferred Foot", "D/W", "A/W",** and **"IR"**.

The **"category"** data type is useful for columns that have a limited set of possible values, such as categorical variables. It can improve performance and reduce memory usage compared to using the **"object"** data type for categorical variables.

Overall, this code is likely being used to optimize the data types of categorical variables in the DataFrame for more efficient processing and memory usage.

#### 2.2 Dealing with inconsistent naming conventions

In [17]:
# Preprocessing the 'LongName' column by converting Unicode characters to ASCII equivalents and identifying abbreviations and acronyms in the dataset
df_copy['LongName'] = df_copy['LongName'].apply(unidecode)
df_copy[df_copy['LongName'].str.contains('\w+\.')]

ID               Name                      LongName  \
1       20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
4      190871          Neymar Jr    Neymar da Silva Santos Jr.   
161    216547               Rafa      Rafael A. Ferreira Silva   
162    200949        Lucas Moura      Lucas Rodrigues M. Silva   
198    187961           Paulinho  Jose Paulo Bezerra M. Junior   
...       ...                ...                           ...   
17840  242941          M. Larsen              Marius B. Larsen   
18345  243428         N. Jenssen            Nicolas P. Jenssen   
18515  242215      D. Norman Jr.              David Norman Jr.   
18517  258527        A. Stenseth        Aleksander B. Stenseth   
18711  258094           K. Ødven           Kristoffer S. Odven   

                                               photoUrl  \
1      https://cdn.sofifa.com/players/020/801/21_60.png   
4      https://cdn.sofifa.com/players/190/871/21_60.png   
161    https://cdn.sofifa.com/players/216/547/21_60.png   
162    https://cdn.sofifa.com/players/200/949/21_60.png   
198    https://cdn.sofifa.com/players/187/961/21_60.png   
...                                                 ...   
17840  https://cdn.sofifa.com/players/242/941/21_60.png   
18345  https://cdn.sofifa.com/players/243/428/21_60.png   
18515  https://cdn.sofifa.com/players/242/215/21_60.png   
18517  https://cdn.sofifa.com/players/258/527/21_60.png   
18711  https://cdn.sofifa.com/players/258/094/21_60.png   

                                               playerUrl Nationality  Age  \
1      http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35   
4      http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28   
161    http://sofifa.com/player/216547/rafael-a-ferre...    Portugal   27   
162    http://sofifa.com/player/200949/lucas-rodrigue...      Brazil   27   
198    http://sofifa.com/player/187961/jose-paulo-bez...      Brazil   31   
...                                                  ...         ...  ...   
17840  http://sofifa.com/player/242941/marius-b-larse...      Norway   20   
18345  http://sofifa.com/player/243428/nicolas-p-jens...      Norway   18   
18515  http://sofifa.com/player/242215/david-norman-j...      Canada   22   
18517  http://sofifa.com/player/258527/aleksander-b-s...      Norway   19   
18711  http://sofifa.com/player/258094/kristoffer-s-o...      Norway   18   

       ↓OVA  POT                            Club  ... SHO PAS  DRI  DEF PHY  \
1        92   92                        Juventus  ...  93  81   89   35  77   
4        91   91             Paris Saint-Germain  ...  85  86   94   36  59   
161      83   83                      SL Benfica  ...  73  74   87   50  52   
162      83   83               Tottenham Hotspur  ...  78  72   87   43  67   
198      83   83  Guangzhou Evergrande Taobao FC  ...  82  77   82   80  85   
...     ...  ...                             ...  ...  ..  ..  ...  ...  ..   
17840    55   69                         Odds BK  ...  49  47   58   30  38   
18345    53   73                  Stabæk Fotball  ...  25  32   37   53  60   
18515    52   62                     Inter Miami  ...  40  51   54   46  61   
18517    52   61                 Strømsgodset IF  ...  51  43   52   25  51   
18711    51   68                    Aalesunds FK  ...  47  51   58   40  41   

       Hits player_transfer_agreement start_year end_year  contract_duration  
1       562                  Contract       2018     2022                  4  
4       595                  Contract       2017     2022                  5  
161     128                  Contract       2016     2024                  8  
162     115                  Contract       2018     2024                  6  
198     124                  Contract       2019     2023                  4  
...     ...                       ...        ...      ...                ...  
17840     0                  Contract       2017     2020                  3  
183

The first line of code is cleaning and preprocessing the column named **"LongName"** in dataset by applying the **"unidecode"** function to the values in the existing **"LongName"** column.

The second line of code is filtering the DataFrame to include only rows where the **"LongName"** column contains at least one word character **(\w)** followed by a period **(.)**. This pattern will match strings that contain an abbreviation or acronym, such as **"Mr."**, **"Ph.D."**, or **"U.S.A."**.

Overall, this lines of code are used to clean and preprocess text data by removing non-ASCII characters and identifying abbreviations and acronyms.

In [18]:
# Taking care of inconsistent naming conventions in the 'LongName' column
name_dict = {
    'C. Ronaldo dos Santos Aveiro': 'Cristiano Ronaldo dos Santos Aveiro',
    'Neymar da Silva Santos Jr.': 'Neymar da Silva Santos Junior',
    'Rafael A. Ferreira Silva': 'Rafael Augusto Ferreira Silva',
    'Lucas Rodrigues M. Silva': 'Lucas Rodrigues Moura da Silva',
    'Jose Paulo Bezerra M. Junior': 'Jose Paulo Bezerra Maciel Junior',
    'Rosberto J. Dourado Santos': 'Rosberto Jose Dourado Santos',
    'Ismaily Goncalves dos S.': 'Ismaily Goncalves dos Santos',
    'Aluisio Chaves Rib. Moraes Junior': 'Aluisio Chaves Ribeiro Moraes Junior',
    'Joao Pedro G. Santos Galvao': 'Joao Pedro Gomes Santos Galvao',
    'Gabriel dos S. Magalhaes': 'Gabriel dos Santos Magalhaes',
    'G. Pierre Paiva Souza': 'Gustavo Pierre Paiva Souza',
    'A. Benjamin Chiamuloira Paes': 'Andre Benjamin Chiamuloira Paes',
    'Wellingt. Edson Sabrao Rolim': 'Wellingtton Edson Sabrao Rolim',
    'Manuel H. Tavares Fernandes': 'Manuel Henrique Tavares Fernandes',
    'Gabriel Vasconcelos F.': 'Gabriel Vasconcelos Ferreira',
    'Francisco L. Lima Silva Machado': 'Francisco Luiz Lima Silva Machado',
    'Adrien S. Perruchet Silva': 'Adrien Sebastien Perruchet Silva',
    'Xabier Etxeita Gorritxat.': 'Xabier Etxeita Gorritxategi',
    'Victorino M. Magela Sa': 'Victorino Martins Magela Sa',
    'Guilherme A. Vieira dos Santos': 'Guilherme Augusto Vieira dos Santos',
    'Juliano Wagner Mascarinhas Concp.': 'Juliano Wagner Mascarenhas Conceicao',
    'Mauro Jaqueson J. Ferreira Santos': 'Mauro Jaqueson Junior Ferreira Santos',
    'F. Evanilson de Lima Barbosa': 'Fernando Evanilson de Lima Barbosa',
    'M. Vanderlei Paulino Kenedy': 'Marcos Vanderlei Paulino Kenedy',
    'Pedro K. Medina da Silva': 'Pedro Kenzo Medina da Silva',
    'Rogerio de A. S. Coutinho': 'Rogerio de Assis Silva Coutinho',
    'Lourency N. Rodrigues': 'Lourency Nascimento Rodrigues',
    'Ivo Daniel F. Mendonca Pinto': 'Ivo Daniel Fernandes Mendonca Pinto',
    'Fernando H. Boldrin': 'Fernando Henrique Boldrin',
    'Carlos Alberto Carvalho Silva Jr.': 'Carlos Alberto Carvalho Silva Junior',
    'Fco. Javier Hidalgo Gomez': 'Francisco Javier Hidalgo Gomez',
    'Adriano Well. Bastos Oliveira': 'Adriano Wellington Bastos Oliveira',
    'Ruben Micael F. da Ressureicao': 'Ruben Micael Fernandes da Ressureicao',
    'Weverson L. de Oliveira Moura': 'Weverson Leandro de Oliveira Moura',
    'Roderick J. Goncalves Miranda': 'Roderick Goncalves Miranda',
    'Fabricio I. Fonseca de Jesus': 'Fabricio Fonseca de Jesus',
    'Hugo Miguel Barreto H. Marques': 'Hugo Miguel Barreto Henrique Marques',
    'Antonaldo V. Laforte Beretta': 'Antonaldo Laforte Beretta',
    'Antonio C. Glauder Garcia': 'Antonio Cristian Glauder Garcia',
    'Martin S. Ellingsen': 'Martin Ellingsen',
    'Rivaldo Vitor Borba Ferreira Jr.': 'Rivaldo Vitor Borba Ferreira Junior',
    'Fco. Javier Atienza Valverde': 'Francisco Javier Atienza Valverde',
    'S. Jefferson Espinho Sobri': 'Sergio Jefferson Espinho Sobri',
    'Admilson E. Dias de Barros': 'Admilson Estaline Dias de Barros',
    'Tiago B. de Melo Tomas': 'Tiago Barreiros de Melo Tomas',
    'Anders K. Jacobsen': 'Anders Kvindebjerg Jacobsen',
    'A.J. DeLaGarza': 'Adolph Joseph DeLaGarza',
    'Felix Alexandre A. Sanches Correia': 'Felix Alexandre Alves Sanches Correia',
    'Andre Filipe M. Ribeiro Ferreira': 'Andre Filipe Martins Ribeiro Ferreira',
    'Angelo Rafael O. Sousa Taveira': 'Angelo Rafael Oliveira Sousa Taveira',
    'Derrick Etienne Jr.': 'Derrick Etienne Junior',
    'David Jose Gomes O. Tavares': 'David Jose Gomes Oliveira Tavares',
    'Erico C. da Silva': 'Erico Constantino da Silva',
    'Ulysses Llanez Jr.': 'Ulysses Llanez Junior',
    'Tommy St. Jago': 'Tommy Jago',
    'Fredrik P. Knudsen': 'Fredrik Pettersen Knudsen',
    'Hugo Gomes dos S. Silva': 'Hugo Gomes dos Santos Silva',
    'Miullen N. Felicio Carva': 'Miullen Nunes Felicio Carvalho',
    'Renan Abner C. de Oliveira': 'Renan Abner Cardoso de Oliveira',
    'Kornelius N. Hansen': 'Kornelius Nyborg Hansen',
    'Kristoffer N. Hansen': 'Kristoffer Nyhus Hansen',
    'Rui F. da Cunha Correia': 'Rui Ferreira da Cunha Correia',
    'Earl Edwards Jr.': 'Earl Edwards Junior',
    'Tiago Fontoura F. Morais': 'Tiago Fontoura Ferreira Morais',
    'Alexander B. Hansen': 'Alexander Bjorneboe Hansen',
    'Dayne St. Clair': 'Dayne Tristan Clair',
    'Magnus S. Lundal': 'Magnus Stian Lundal',
    'Elias K. Hagen': 'Elias Kristoffer Hagen',
    'Sebastian S. Sebulonsen': 'Sebastian Stian Sebulonsen',
    'Magnus R. Jensen': 'Magnus Rudolf Jensen',
    'Marius B. Larsen': 'Marius Bjornerud Larsen',
    'Nicolas P. Jenssen': 'Nicolas Pedersen Jenssen',
    'David Norman Jr.': 'David Norman Junior',
    'Aleksander B. Stenseth': 'Aleksander Bjornevik Stenseth',
    'Kristoffer S. Odven': 'Kristoffer Storflor Odven'}

df_copy['LongName'] = df_copy['LongName'].replace(name_dict)
df_copy.loc[df_copy['ID'] == 226853,'LongName'] = 'Jeremiah Israël Juste'

The **"replace"** method in pandas allows you to replace values in a DataFrame with other values. In this case, the method is being called on the **"LongName"** column of the DataFrame, and the replacement values are taken from the **"name_dict"** dictionary.

The **"name_dict"** dictionary contains mappings between original names and their updated or corrected versions. This code can be useful for standardizing names in a dataset, such as replacing abbreviations or nicknames with full names or correcting spelling errors.

In [19]:
# Getting all the column names
df_copy.columns

Index(['ID', 'Name', 'LongName', 'photoUrl', 'playerUrl', 'Nationality', 'Age',
       '↓OVA', 'POT', 'Club', 'Contract', 'Positions', 'Height', 'Weight',
       'Preferred Foot', 'BOV', 'Best Position', 'Joined', 'Loan Date End',
       'Value', 'Wage', 'Release Clause', 'Attacking', 'Crossing', 'Finishing',
       'Heading Accuracy', 'Short Passing', 'Volleys', 'Skill', 'Dribbling',
       'Curve', 'FK Accuracy', 'Long Passing', 'Ball Control', 'Movement',
       'Acceleration', 'Sprint Speed', 'Agility', 'Reactions', 'Balance',
       'Power', 'Shot Power', 'Jumping', 'Stamina', 'Strength', 'Long Shots',
       'Mentality', 'Aggression', 'Interceptions', 'Positioning', 'Vision',
       'Penalties', 'Composure', 'Defending', 'Marking', 'Standing Tackle',
       'Sliding Tackle', 'Goalkeeping', 'GK Diving', 'GK Handling',
       'GK Kicking', 'GK Positioning', 'GK Reflexes', 'Total Stats',
       'Base Stats', 'W/F', 'SM', 'A/W', 'D/W', 'IR', 'PAC', 'SHO', 'PAS',
       'DRI', 'DEF', 

In [20]:
# Taking care of inconsistent naming conventions in the column names in this dataset
new_names = {
    'ID': 'id',
    'LongName': 'player_full_name',
    'Nationality': 'nationality',
    'Age': 'age',
    '↓OVA': 'overall_rating',
    'POT': 'potential_rating',
    'Club': 'club',
    'Contract': 'contract',
    'Positions': 'positions',
    'Height': 'height_cm',
    'Weight': 'weight_kg',
    'Preferred Foot': 'preferred_foot',
    'BOV': 'best_overall_rating',
    'Best Position': 'best_position',
    'Joined': 'joined_date',
    'Loan Date End': 'expiry_date',
    'Value': 'value_euro',
    'Wage': 'wage_euro',
    'Release Clause': 'release_clause_euro',
    'Attacking': 'attacking_rating',
    'Crossing': 'crossing_rating',
    'Finishing': 'finishing_rating',
    'Heading Accuracy': 'heading_accuracy_rating',
    'Short Passing': 'short_passing_rating',
    'Volleys': 'volleys_rating',
    'Skill': 'skill_rating',
    'Dribbling': 'dribbling_rating',
    'Curve': 'curve_rating',
    'FK Accuracy': 'free_kick_accuracy_rating',
    'Long Passing': 'long_passing_rating',
    'Ball Control': 'ball_control_rating',
    'Movement': 'movement_rating',
    'Acceleration': 'acceleration_rating',
    'Sprint Speed': 'sprint_speed_rating',
    'Agility': 'agility_rating',
    'Reactions': 'reactions_rating',
    'Balance': 'balance_rating',
    'Power': 'power_rating',
    'Shot Power': 'shot_power_rating',
    'Jumping': 'jumping_rating',
    'Stamina': 'stamina_rating',
    'Strength': 'strength_rating',
    'Long Shots': 'long_shots_rating',
    'Mentality': 'mentality_rating',
    'Aggression': 'aggression_rating',
    'Interceptions': 'interceptions_rating',
    'Positioning': 'positioning_rating',
    'Vision': 'vision_rating',
    'Penalties': 'penalties_rating',
    'Composure': 'composure_rating',
    'Defending': 'defending_rating',
    'Marking': 'marking_rating',
    'Standing Tackle': 'standing_tackle_rating',
    'Sliding Tackle': 'sliding_tackle_rating',
    'Goalkeeping': 'goalkeeping_rating',
    'GK Diving': 'gk_diving_rating',
    'GK Handling': 'gk_handling_rating',
    'GK Kicking': 'gk_kicking_rating',
    'GK Positioning': 'gk_positioning_rating',
    'GK Reflexes': 'gk_reflexes_rating',
    'Total Stats': 'total_stats',
    'Base Stats': 'base_stats',
    'W/F': 'weak_foot',
    'SM': 'skill_moves',
    'A/W': 'attacking_work_rate',
    'D/W': 'defensive_work_rate',
    'IR': 'international_reputation',
    'PAC': 'pace_rating',
    'SHO': 'shooting_rating',
    'PAS': 'passing_rating',
    'DRI': 'dribbling_rating',
    'DEF': 'defending_rating',
    'PHY': 'physical_rating',
    'Hits': 'no_of_profile_hits'
}
df_copy = df_copy.rename(columns=new_names)

First, a dictionary of old and new column names is created and assigned to the variable **new_names**. Then, the **rename()** method is called on the DataFrame object **df_copy**, with the columns parameter set to the new_names dictionary. This renames the columns of the DataFrame to the new names specified in the dictionary.

Overall, this code provides an easy and efficient way to rename the columns of a pandas DataFrame in Python, using the **rename()** method and a dictionary of new column names.

#### 2.3 Removing redundant data

In [21]:
# Removing irrelevant columns
df_copy = df_copy.drop(['start_year','end_year','contract','positions','playerUrl','photoUrl','Name'],axis=1)

This code provides an efficient way to drop multiple columns from the DataFrame using the **drop()** method.

### 3. Conclusion

The cleaning of this dataset has helped us to eliminate errors and inconsistencies, making it more reliable and accurate for further analysis. The cleaned dataset provides a detailed account of the football players. By organizing and structuring the data in a more presentable format, it is now easier to identify patterns and trends in the performance and career progression of these players. The cleaned dataset can be used for research purposes or to build predictive models that can help in making informed decisions in the football gaming industry.

In [22]:
# Final look on the cleaned dataset before exporting
df_copy.head()

id                     player_full_name nationality  age  \
0  158023                         Lionel Messi   Argentina   33   
1   20801  Cristiano Ronaldo dos Santos Aveiro    Portugal   35   
2  200389                            Jan Oblak    Slovenia   27   
3  192985                      Kevin De Bruyne     Belgium   29   
4  190871        Neymar da Silva Santos Junior      Brazil   28   

   overall_rating  potential_rating                 club  height_cm  \
0              93                93         FC Barcelona        170   
1              92                92             Juventus        187   
2              91                93      Atlético Madrid        188   
3              91                91      Manchester City        181   
4              91                91  Paris Saint-Germain        175   

   weight_kg preferred_foot  ...  international_reputation pace_rating  \
0       72.0           Left  ...                        5           85   
1       83.0          Right  ...                        5           89   
2       87.0          Right  ...                        3           87   
3       70.0          Right  ...                        4           76   
4       68.0          Right  ...                        5           91   

  shooting_rating passing_rating  dribbling_rating  defending_rating  \
0              92             91                95                38   
1              93             81                89                35   
2              92             78                90                52   
3              86             93                88                64   
4              85             86                94                36   

   physical_rating  no_of_profile_hits  player_transfer_agreement  \
0               65                 771                   Contract   
1               77                 562                   Contract   
2               90                 150                   Contract   
3               78                 207                   Contract   
4               59                 595                   Contract   

   contract_duration  
0                 17  
1                  4  
2                  9  
3                  8  
4                  5  

[5 rows x 74 columns]

In [23]:
# Exporting the clean data
df_copy.to_csv('./data/fifa_21_clean.csv')